Refuse Enrollments if the public key is already present

linked0 opened the pull request:
We should refuse any Enrollment which would lead to the same public key being enrolled twice (despite the two UTXOs). In other words, enrollments from the same public key can not be accepted in an overlapped cycles. But, a validator can enroll with any UTXO after a validator cycle ends.

This PR is for that requirement. This contains network tests also.

Fixes #1310

linked0 committed:
Network test for enrolling with same public key

linked0 committed:
Refuse Enrollments if public key is present

We should refuse any Enrollment which would lead to the same public key
being enrolled twice (despite the two UTXOs). But it's possible to
enroll those enrollments without their cycles overlaping.

linked0 committed:
Refactor code about restoring validator set

There was two points in the `Ledger` where `EnrollmentManager`'s
addValidators is called, which was error-prone. We can just use
`Ledger`'s `updateValidatorSet` function for adding new validators.
Its use can make code maintenance simpler.

linked0 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -94,15 +95,20 @@ public class ValidatorSet
             return "This validator is already enrolled";
 
+        // check if an enrollment of the same public key is already present
+        if (this.hasPublicKey(pubkey))
```
This checking code should be also called in the `isInvalidCandidateReason` function of `EnrollmentManager`. I will fix it.

linked0 commented:
After updating some code, I will request a review for it again.

AndrejMitrovic commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -237,6 +243,34 @@ public class ValidatorSet
+        }
+        catch (Exception ex)
+        {
+            log.error("Exception occured on hasEnrollment: {}, " ~
```
We need to stop writing code like this.

If there is an error in the database, then the enrollment will be accepted. Do you really want that to happen?

linked0 replied:
> If there is an error in the database, then the enrollment will be accepted. Do you really want that to happen?

I added `assert(0)` statements into the two critical functions, `hasEnrollment` and `hasPublicKey`. But we would need to check more about other functions as a separate issue.

AndrejMitrovic replied:
Thanks. There is still an open issue about handling SQLite errors properly. Hopefully someone takes care of that.

linked0 committed:
Call assert(0) when database does not work

The code checking the enrollment already exists should work without
abnormal operation of the datbase. So we add the `assert(0)` call into
those critical function. We need to check all the function of this kind.

linked0 commented:
Ready for review

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -200,8 +201,9 @@ public class EnrollmentManager
 
-    public bool addEnrollment (const ref Enrollment enroll, Height block_height,
-        scope UTXOFinder finder) @safe nothrow
+    public bool addEnrollment (const ref Enrollment enroll,
+        const ref PublicKey pubkey, Height block_height, scope UTXOFinder finder)
+        @safe nothrow
```
Why do we need to pass the public key if we look up the UTXO ?

Geod24 replied:
Okay, looking at the rest of the code, it might make sense to just pass the `UTXO` and remove the `finder` and `pubkey` argument.

linked0 replied:
OK, I understand you meant we can remove also the `finder` argument on `isInvalidReason`. Right?
In fact, the `finder` is used for `isInvalidReason`  in  `EnrollmentPool`'s `add` function.

Geod24 replied:
Ah, I see, so it will be a bit complicated. Let's leave it for another day then.

Geod24 commented:
First commit is great. Second and third should be merged, and note the comment about UTXO. Last commit, I'm not so sure about it. I don't think we need it right away anyway, do we ?

Geod24 commented:
Gave it a second read and realized I missed Drey's comment. Let's go ahead with this, thanks!
